### PR TITLE
[ehancement](stats) Stats preheating as FE booted

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

1. Support prefetch some column stats when FE booted, it would load column stats that was got updated recently according to the comment of PR #18460 from @morrySnow  
2. Refactor stats cache, split histogram cache from column stats, so that we could avoid some redundant query for column statistics table,for example, update the histogram or column stats only, in the previous implementation a united cache loader would send query request to both column stats table and histogram table, 
3. Extract some common logic to StatsUtil
4. Remove some useless codes in unit tests, those codes is hard to maintaince and it's not a good idea for testing the accurracy of stats estimation according to the advise from @englefly 
5. Add field type restriction when create analysis tasks to avoid unnecessary failure
## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatistic.java [hunk 1]
```diff
@@ -154,7 +154,8 @@ public static ColumnStatistic fromResultRow(ResultRow resultRow) {
             }
             columnStatisticBuilder.setSelectivity(1.0);
             columnStatisticBuilder.setOriginalNdv(ndv);
-            Histogram histogram = Env.getCurrentEnv().getStatisticsCache().getHistogram(tblId, idxId, colName);
+            Histogram histogram = Env.getCurrentEnv().getStatisticsCache().getHistogram(tblId, idxId, colName)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatistic.java:158]: maybe the better way is, one class use for caching column stats, one class use for caching histogram, and a new class use for stats derive, it contains both column stats and histogram

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java [hunk 2]
```diff
@@ -44,85 +47,155 @@ public class StatisticsCache {
...
+        columnStatisticsCache.synchronous().invalidate(new StatisticsCacheKey(tblId, idxId, colName));
     }
 
-    public void updateCache(long tblId, long idxId, String colName, ColumnLevelStatisticCache statistic) {
-        cache.synchronous().put(new StatisticsCacheKey(tblId, idxId, colName), statistic);
+    public void updateCache(long tblId, long idxId, String colName, ColumnStatistic statistic) {
+        columnStatisticsCache.synchronous().put(new StatisticsCacheKey(tblId, idxId, colName), Optional.of(statistic));
     }
 
     public void refreshSync(long tblId, long idxId, String colName) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java]: histogram update and refresh should not call these functions

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java [hunk 3]
```diff
@@ -44,85 +47,155 @@ public class StatisticsCache {
...
-        cache.synchronous().refresh(new StatisticsCacheKey(tblId, idxId, colName));
+        columnStatisticsCache.synchronous().refresh(new StatisticsCacheKey(tblId, idxId, colName));
     }
+
+    public void preHeat() {
+        threadPool.submit(this::doPreHeat);
+    }
+
+    private void doPreHeat() {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java:147]: does preheat load histogram?

File: fe/fe-core/src/test/java/org/apache/doris/nereids/jobs/cascades/DeriveStatsJobTest.java [hunk 4]
```diff
@@ -70,7 +70,7 @@ public void testExecute() throws Exception {
         }
         Statistics statistics = cascadesContext.getMemo().getRoot().getStatistics();
         Assertions.assertNotNull(statistics);
-        Assertions.assertTrue(Precision.equals(0.5, statistics.getRowCount(), 0.1));
+        Assertions.assertTrue(Precision.equals(0, statistics.getRowCount(), 0.1));
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/jobs/cascades/DeriveStatsJobTest.java:73]: why this test changed?

File: fe/fe-core/src/test/java/org/apache/doris/nereids/stats/TPCHStats.java
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/stats/TPCHStats.java:1]: why remove this test?

File: fe/fe-core/src/test/java/org/apache/doris/statistics/CacheTest.java [hunk 5]
```diff
@@ -40,7 +40,7 @@
 public class CacheTest extends TestWithFeService {
 
     @Test
-    public void testColumn(@Mocked StatisticsCacheLoader cacheLoader) throws Exception {
+    public void testColumn(@Mocked ColumnStatisticsCacheLoader cacheLoader) throws Exception {
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/statistics/CacheTest.java:43]: does need test histogram cache?

**Comment by Kikyou1997** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisTaskScheduler.java [hunk 6]
```diff
@@ -47,25 +44,24 @@ public class AnalysisTaskScheduler {
...
-            systemJobQueue.remove(analysisJobInfo);
-            manualJobSet.remove(analysisJobInfo);
+    public synchronized void schedule(AnalysisTaskInfo analysisTaskInfo) {
+        try {
+            TableIf table = StatisticsUtil.findTable(analysisTaskInfo.catalogName,
+                    analysisTaskInfo.dbName, analysisTaskInfo.tblName);
+            BaseAnalysisTask analysisTask = table.createAnalysisTask(this, analysisTaskInfo);
+            switch (analysisTaskInfo.jobType) {
+                case MANUAL:
+                    addToManualJobQueue(analysisTask);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisTaskScheduler.java:54]: add unsafe perfix to all unsafe function, such as rename `addToManualJobQueue ` to `unsafeAddToManualJobQueue`. Or add some comment to all unsafe function

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java [hunk 7]
```diff
@@ -44,85 +47,159 @@ public class StatisticsCache {
...
 
     // TODO: finish this method.
     public void eraseExpiredCache(long tblId, long idxId, String colName) {
-        cache.synchronous().invalidate(new StatisticsCacheKey(tblId, idxId, colName));
+        columnStatisticsCache.synchronous().invalidate(new StatisticsCacheKey(tblId, idxId, colName));
     }
 
-    public void updateCache(long tblId, long idxId, String colName, ColumnLevelStatisticCache statistic) {
-        cache.synchronous().put(new StatisticsCacheKey(tblId, idxId, colName), statistic);
+    public void updateCache(long tblId, long idxId, String colName, ColumnStatistic statistic) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java:131]:
```suggestion
    public void updateColStatsCache(long tblId, long idxId, String colName, ColumnStatistic statistic) {
```

File: fe/fe-core/src/test/java/org/apache/doris/nereids/jobs/cascades/DeriveStatsJobTest.java (hunk 4 above)
Comment (Kikyou1997) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/jobs/cascades/DeriveStatsJobTest.java:73]: Remove some useless codes in unit tests, those codes is hard to maintaince and it's not a good idea for testing the accurracy of stats estimation according to the advise from @englefly

File: fe/fe-core/src/test/java/org/apache/doris/nereids/stats/TPCHStats.java
Comment (Kikyou1997) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/stats/TPCHStats.java:1]: Remove some useless codes in unit tests, those codes is hard to maintaince and it's not a good idea for testing the accurracy of stats estimation according to the advise from @englefly

File: fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatistic.java (hunk 1 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/ColumnStatistic.java:158]: There is not much difference between these two implementations. I might make the changes as you suggested when it becomes necessary to do so.

**Comment by Kikyou1997** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java (hunk 3 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsCache.java:147]: no

**Review by englefly — APPROVED** [review]

**Pull request merged**